### PR TITLE
Improve val step logging

### LINKED ISSUE #7346: [Bug] Logging Separately Per Epoch
## 🐛 Bug

Logs separate plots for each epoch

### To Reproduce

```python
import os
import torch
from torch.utils.data import Dataset, DataLoader
from pytorch_lightning import LightningModule, Trainer
from torch.nn import functional as F
from torchmetrics.classification import Accuracy

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)
    def __getitem__(self, index):
        return self.data[index], 1
    def __len__(self):
        return self.len

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.loss_fn = torch.nn.CrossEntropyLoss()
        self.train_accuracy = Accuracy()
        self.val_accuracy = Accuracy()

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_id):
        return self.step(batch, step_type="train")

    def validation_step(self, batch, batch_id):
        return self.step(batch, step_type="val")

    def step(self, batch, step_type: str):
        data, y = batch
        y_pred = self(data)
        loss = self.loss_fn(y_pred, y)
        self.log(f"{step_type}_loss", loss, prog_bar=True, sync_dist=False)
        # Compute accuracy
        y_pred = F.softmax(y_pred, dim=1)
        self.__getattr__(f"{step_type}_accuracy")(y_pred, y)
        self.log(f"{step_type}_accuracy", self.__getattr__(f"{step_type}_accuracy"), on_step=True, on_epoch=True)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=5,
        limit_val_batches=5,
        num_sanity_val_steps=0,
        max_epochs=4,
        weights_summary=None,
    )
    trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    trainer.test(model, test_dataloaders=test_data)

if __name__ == '__main__':
    run()
```

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #7346 

- Changes val step metrics to no longer append `/epoch_*` to the metric names (which creates many plots in e.g. tensorboard)
- Instead they now store a local step for val and test. Not ideal but this will be changed again in upcoming logger connector refactors (cc @carmocca)
- Refactors `log_evaluation_step_metrics` to be in the `LoggerConnector` (where it belongs haha)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [N/A] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -371,3 +374,31 @@ def log_train_step_metrics(self, batch_output):
             if len(batch_log_metrics) > 0 or len(grad_norm_dic) > 0:
                 self.log_metrics(batch_log_metrics, grad_norm_dic)
                 self._callback_metrics.update(batch_log_metrics)
+
+    @property
+    def evaluation_log_step(self) -> Optional[int]:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:379]: why optional?

Comment (ethanwharris) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:379]: Just because running stage can be something other than validating or testing, so it's possible for the method to return None (though not the way we use it)

Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:379]: could we make it more clear by adding
```
else:
    return None
```
at the end?

**Review by SkafteNicki — APPROVED** [review]

**Comment by awaelchli** [comment]
image

Red: master
Blue: this branch

You mention this will change again in a later "refactor". What will change?

**Comment by ethanwharris** [comment]
@awaelchli I think the functionality will stay the same, only the code will change. The screenshot you show is correct and gives the new behaviour following this PR. Val step logging is and odd usage anyway but at least this stops the potentially hundreds of graphs from being created in tensorboard.

**Comment by awaelchli** [comment]
Okay, as long as all the other logging for training and epoch end is not affected. Consider changing the labelling here because these per-epoch plots being created was a design decision and therefore expected behaviour.

**Comment by awaelchli** [comment]
> Val step logging is and odd usage anyway 

could not agree more 🤣

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -371,3 +374,31 @@ def log_train_step_metrics(self, batch_output):
...
+        elif self.trainer.state.stage is RunningStage.TESTING:
+            return self._test_log_step
+
+    def increment_evaluation_log_step(self) -> None:
+        if self.trainer.state.stage is RunningStage.VALIDATING:
+            self._val_log_step += 1
+        elif self.trainer.state.stage is RunningStage.TESTING:
+            self._test_log_step += 1
+
+    def log_evaluation_step_metrics(self):
```
Comment (edgarriba) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:391]:
```suggestion
    def log_evaluation_step_metrics(self) -> None:
```

**Review by edgarriba — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 3]
```diff
@@ -371,3 +374,31 @@ def log_train_step_metrics(self, batch_output):
...
+        if self.trainer.sanity_checking:
+            return
+        _, batch_log_metrics = self.cached_results.update_logger_connector()
+
+        # logs user requested information to logger
+        if len(batch_log_metrics) > 0:
+            if "step" in batch_log_metrics:
+                self.log_metrics(batch_log_metrics, {})
+            else:
+                self.log_metrics(batch_log_metrics, {}, step=self.evaluation_log_step)
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:401]:
```suggestion
            kwargs = dict() if "step" in batch_log_metrics else dict(step=self.evaluation_log_step)
            self.log_metrics(batch_log_metrics, {}, **kwargs)
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 4]
```diff
@@ -371,3 +374,29 @@ def log_train_step_metrics(self, batch_output):
             if len(batch_log_metrics) > 0 or len(grad_norm_dic) > 0:
                 self.log_metrics(batch_log_metrics, grad_norm_dic)
                 self._callback_metrics.update(batch_log_metrics)
+
+    @property
+    def evaluation_log_step(self) -> Optional[int]:
+        if self.trainer.state.stage is RunningStage.VALIDATING:
+            return self._val_log_step
+        elif self.trainer.state.stage is RunningStage.TESTING:
+            return self._test_log_step
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:383]:
```suggestion
    @property
    def evaluation_log_step(self) -> Optional[int]:
        if self.trainer.state.stage is RunningStage.VALIDATING:
            return self._val_log_step
        elif self.trainer.state.stage is RunningStage.TESTING:
            return self._test_log_step
        else:
            return None
```

**Review by ananthsub — APPROVED** [review]
